### DISCUSSION TITLE
How to extract_image CMYK to RGB with no resampling ?

### DISCUSSION THREAD

**Discussion by denis-bz** [pymupdf/PyMuPDF#2104] [Q&A]
This is more likely my lack of understanding than an issue:
I have a .pdf with a CMYK image, and want an RGB.png with no resampling.
Is there a recipe for that ?

Following https://pymupdf.readthedocs.io/en/latest/document.html#Document.extract_image
```
d = doc.extract_image(xref)
print( d["height"], d["width"], len(d["image"]) )  # 688 660 46727  hmm, compressed ?
```

(The top answer to [SO extract-images-from-pdf-without-resampling-in-python](https://stackoverflow.com/questions/2693820/extract-images-from-pdf-without-resampling-in-python)
version for fitz 1.19.6 calls `extract_image` but doesn't use it ?!
```
xref = img[0]
image = doc.extract_image(xref)  # <-- not used ?
pix = fitz.Pixmap(doc, xref)
pix.save( ... )
```

My pdf is pretty big to post, is there a small testcase ?
Thanks,
cheers

**Comment by JorjMcKie** [comment]
Converted your post to a Discussion item right away - clearly not an issue.
* You can **_either_** use `doc.extract_image(xref)` in which case the image is extracted as-is - at least when MuPDF was able to recognize its format.
* **_Or_** extract via `fitz.Pixmap(doc. xref)`. Not surprisingly, in this case MuPDF interprets the image data so they fit in a rectangular scheme of width * height (* n for the pixel size). The original image format is unkown at that point and will neither be recreatable.

In your case, bullet point 1 obviously happens. So you should be given the original image. How to convert this animal to a different image format (namely RGB) **_without resampling_** is a question that I cannot answer and which is outside PyMuPDF's scope.
I am not even sure if your request is possible at all.

If you use the second method, you will get a Pixmap in CMYK format alright, then you can convert it to RGB by creating another pixmap, naming a different (i.e. RGB) colorspace:
```python
pix1 = fitz.Pixmap(doc, xref)
pix2 = fitz.Pixmap(fitz.csRGB, pix1)
```
Same thing happens using this approach:
```python
img = doc.extract_image(xref)
pix1 = fitz.Pixmap(img["image"])  # detects format from passed in bytes
pix2 = fitz.Pixmap(fitz.csRGB, pix1)
```

Of course you can also use packages like Pillow to make an image from the bytes object ``img["image"]`` and convert this one to RGB.

**Reply by denis-bz** [reply]
> Of course you can also use packages like Pillow to make an image from the bytes object img["image"]

trouble is that "image" is too short -- notice the line
```
print( d["height"], d["width"], len(d["image"]) )  # 688 660 46727  hmm, compressed ?

# push on regardless ?
b = PIL.Image.frombytes( "CMYK", (w, h), image )  # ValueError: not enough image data

pix = fitz.Pixmap( fitz.csCMYK, image )
print( f"Pixmap fitz.csCMYK:  h {pix.height}  w {pix.width}  {pix.colorspace} " )
    # Pixmap fitz.csCMYK:  h 0  w 0  Colorspace(CS_CMYK) - DeviceCMYK
b = pix.tobytes( "png" )  # ValueError: unsupported colorspace for 'png'
pix.pixel( 0, 0 )  # ValueError: pixel(s) outside image
pix.color_count()  # Fatal Python error: Segmentation fault
```
Would you have a pdf with a CMYK image for which `extract_image` works, i.e. gives h x w x 4 uint8 s ?

You're right that CMYK to RGB is a separate issue. Google -> e.g.
https://www.ginifab.com/feeds/pms/cmyk_to_rgb.php
Red = 255 × ( 1 - Cyan ÷ 100 ) × ( 1 - Black ÷ 100 )
Green = 255 × ( 1 - Magenta ÷ 100 ) × ( 1 - Black ÷ 100 )
Blue = 255 × ( 1 - Yellow ÷ 100 ) × ( 1 - Black ÷ 100 )
This may not be color-accurate, don't know.

**Reply by JorjMcKie** [reply]
why dont you simply open the image from the bytes directly and let Pillow do the analysis? E.g.
```python
fp=io.BytesIO(img["image"])
img=Image(fp)
from PIL import Image
img=Image.open(fp)
img.info
{'jfif': 257, 'jfif_version': (1, 1), 'dpi': (96, 96), 'jfif_unit': 1, 'jfif_density': (96, 96), 'progressive': 1, 'progression': 1}
```

**Reply by denis-bz** [reply]
Sure, but this "image" is too short, looks buggy: `print( d["height"], d["width"], len(d["image"]) )  # 688 660 46727`.

Don't you think a little CMYK  testcase would be useful ?

**Reply by JorjMcKie** [reply]
What does Pillow say if you do the above?

**Reply by denis-bz** [reply]
Thanks, that at least generates h x w x 4 --
```
fp = io.BytesIO(d["image"])
img = Image.open(fp)  # PIL.JpegImagePlugin.JpegImageFile
print( img.info )  # {'adobe': 100, 'adobe_transform': 2}

img.save( "tmp.jpg" )
cmyk = np.asarray( Image.open( "tmp.jpg" ).getdata() ).reshape( h, w, 4 )
counts( cmyk )  # 256 256 256 95
```

will convert to RGB and let you know, thanks.

Re a little testcase: must be loads of such ... someplace ?

**Comment by JorjMcKie** [comment]
You can look at the PDF xref of the image and `print(doc.xref_object(xref)`. Will show also length parameters.
There also is the chance of a damaged PDF, you know.
